File: microservices/auth_service/utils/jwt_util.py
"""jwt_util.py – JWT generation, validation, and role-encoding utilities.

Tasks: 3 (JWT generation/validation) + 6 (RBAC role encoding in JWT).
"""

from __future__ import annotations

import os
import uuid
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

import jwt

# ---------------------------------------------------------------------------
# Mock role data (Task 6 – role retrieval assumed handled by mock/fixed data)
# ---------------------------------------------------------------------------

_MOCK_USER_ROLES: Dict[str, List[str]] = {
    "admin": ["ROLE_ADMIN", "ROLE_USER"],
    "alice": ["ROLE_USER"],
    "bob": ["ROLE_USER"],
}

_MOCK_USER_PERMISSIONS: Dict[str, List[str]] = {
    "admin": ["read", "write", "delete"],
    "alice": ["read", "write"],
    "bob": ["read"],
}


def get_user_roles(username: str) -> List[str]:
    """Return roles for *username* from mock data.  Returns empty list if unknown."""
    return _MOCK_USER_ROLES.get(username, ["ROLE_USER"])


def get_user_permissions(username: str) -> List[str]:
    """Return permissions for *username* from mock data."""
    return _MOCK_USER_PERMISSIONS.get(username, ["read"])


# ---------------------------------------------------------------------------
# JWT configuration helpers
# ---------------------------------------------------------------------------


def _get_secret() -> str:
    return os.environ.get("JWT_SECRET_KEY", "jwt-secret-change-in-production")


def _get_algorithm() -> str:
    return os.environ.get("JWT_ALGORITHM", "HS256")


def _get_access_expiry_minutes() -> int:
    return int(os.environ.get("JWT_ACCESS_TOKEN_EXPIRES_MINUTES", "60"))


def _get_refresh_expiry_days() -> int:
    return int(os.environ.get("JWT_REFRESH_TOKEN_EXPIRES_DAYS", "7"))


# ---------------------------------------------------------------------------
# Token generation (Tasks 3 + 6)
# ---------------------------------------------------------------------------


def generate_token(
    user_id: str,
    username: str,
    roles: Optional[List[str]] = None,
    permissions: Optional[List[str]] = None,
    expiry_minutes: Optional[int] = None,
) -> str:
    """Generate a signed JWT containing userId, roles, permissions, and expiry.

    Acceptance criteria satisfied:
    - Task 1/3/6: token includes ``userId``, ``roles``, and ``expiry``.
    - Task 6: roles and permissions are embedded in the payload.

    Parameters
    ----------
    user_id:        Unique identifier for the user.
    username:       Login handle (used to look up roles when not provided).
    roles:          Explicit role list; resolved from mock data when ``None``.
    permissions:    Explicit permission list; resolved from mock data when ``None``.
    expiry_minutes: Override token lifetime in minutes.
    """
    if roles is None:
        roles = get_user_roles(username)
    if permissions is None:
        permissions = get_user_permissions(username)

    now = datetime.now(tz=timezone.utc)
    minutes = expiry_minutes if expiry_minutes is not None else _get_access_expiry_minutes()
    from datetime import timedelta

    expires_at = now + timedelta(minutes=minutes)

    payload: Dict[str, Any] = {
        "userId": user_id,
        "username": username,
        "roles": roles,
        "permissions": permissions,
        "iat": now,
        "exp": expires_at,
        "expiry": expires_at.isoformat(),
        "jti": str(uuid.uuid4()),
        "token_type": "access",
    }

    return jwt.encode(payload, _get_secret(), algorithm=_get_algorithm())


def generate_refresh_token(user_id: str, username: str) -> str:
    """Generate a refresh token with a longer expiry.

    Used by Task 2 (session refresh) and Task 3 (/refresh-token endpoint).
    """
    now = datetime.now(tz=timezone.utc)
    from datetime import timedelta

    expires_at = now + timedelta(days=_get_refresh_expiry_days())

    payload: Dict[str, Any] = {
        "userId": user_id,
        "username": username,
        "iat": now,
        "exp": expires_at,
        "expiry": expires_at.isoformat(),
        "jti": str(uuid.uuid4()),
        "token_type": "refresh",
    }

    return jwt.encode(payload, _get_secret(), algorithm=_get_algorithm())


# ---------------------------------------------------------------------------
# Token validation (Tasks 1 + 3)
# ---------------------------------------------------------------------------


def validate_token(token: str) -> Dict[str, Any]:
    """Decode and validate *token*.

    Returns the decoded payload dict.

    Raises
    ------
    jwt.ExpiredSignatureError  – when the token has expired.
    jwt.InvalidTokenError      – for any other validation failure (wrong format,
                                  bad signature, missing claims, etc.).
    """
    try:
        payload = jwt.decode(
            token,
            _get_secret(),
            algorithms=[_get_algorithm()],
        )
    except jwt.ExpiredSignatureError:
        raise
    except jwt.DecodeError as exc:
        raise jwt.InvalidTokenError(f"Token decode failed: {exc}") from exc
    except jwt.InvalidTokenError:
        raise

    # Reject tokens with incorrect payload formats (Task 1 AC3)
    if "userId" not in payload and "sub" not in payload:
        raise jwt.InvalidTokenError("Token payload missing required userId/sub claim")

    return payload


def renew_token(token: str) -> str:
    """Renew an access token.

    Accepts an expired OR valid token and issues a fresh access token
    preserving the user identity and roles.  Satisfies Task 1 AC2:
    "Expired tokens are successfully renewed."

    Raises
    ------
    jwt.InvalidTokenError – if the token is structurally invalid (cannot be
                            decoded at all, even ignoring expiry).
    """
    try:
        payload = jwt.decode(
            token,
            _get_secret(),
            algorithms=[_get_algorithm()],
            options={"verify_exp": False},
        )
    except jwt.InvalidTokenError as exc:
        raise jwt.InvalidTokenError(f"Cannot renew invalid token: {exc}") from exc

    user_id = payload.get("userId") or payload.get("sub", "")
    username = payload.get("username", "")
    roles = payload.get("roles")
    permissions = payload.get("permissions")

    return generate_token(user_id, username, roles=roles, permissions=permissions)


# ---------------------------------------------------------------------------
# Role-based access checks (Tasks 5 + 6)
# ---------------------------------------------------------------------------


def has_role(token: str, required_role: str) -> bool:
    """Return True if *token* carries *required_role*.

    Task 6 AC2: "Protected endpoint access respects embedded user roles."
    """
    try:
        payload = validate_token(token)
    except jwt.InvalidTokenError:
        return False
    roles: List[str] = payload.get("roles", [])
    return required_role in roles


def check_roles(payload: Dict[str, Any], required_role: str) -> bool:
    """Return True if decoded *payload* includes *required_role*."""
    return required_role in payload.get("roles", [])

File: src/main/java/com/zcloud/platform/util/JwtUtil.java
package com.zcloud.platform.util;

import io.jsonwebtoken.*;
import io.jsonwebtoken.security.Keys;

import java.nio.charset.StandardCharsets;
import java.security.Key;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * JwtUtil – manages JWT token generation, renewal, and validation.
 *
 * <p>Task 1 acceptance criteria:
 * <ul>
 *   <li>AC1: Valid JWT token includes {@code userId}, {@code roles}, and {@code expiry}.</li>
 *   <li>AC2: Expired tokens are successfully renewed.</li>
 *   <li>AC3: Tokens with incorrect payloads return 401 error (throws {@link JwtValidationException}).</li>
 * </ul>
 */
public class JwtUtil {

    /** Default token lifetime in milliseconds (60 minutes). */
    private static final long DEFAULT_EXPIRY_MS = 60L * 60 * 1000;

    private final Key signingKey;
    private final long expiryMs;

    public JwtUtil(String secret) {
        this(secret, DEFAULT_EXPIRY_MS);
    }

    public JwtUtil(String secret, long expiryMs) {
        this.signingKey = Keys.hmacShaKeyFor(secret.getBytes(StandardCharsets.UTF_8));
        this.expiryMs = expiryMs;
    }

    // -----------------------------------------------------------------------
    // Token generation (Task 1 AC1)
    // -----------------------------------------------------------------------

    /**
     * Generate a signed JWT for the given user.
     *
     * <p>The token payload includes:
     * <ul>
     *   <li>{@code userId} – the user's unique identifier</li>
     *   <li>{@code roles}  – the user's roles list</li>
     *   <li>{@code expiry} – ISO-8601 expiry timestamp</li>
     *   <li>{@code jti}    – unique token ID for revocation</li>
     * </ul>
     *
     * @param userId unique user identifier
     * @param roles  list of role strings
     * @return signed JWT string
     */
    public String generateToken(String userId, List<String> roles) {
        Date now = new Date();
        Date expiry = new Date(now.getTime() + expiryMs);

        return Jwts.builder()
                .claim("userId", userId)
                .claim("roles", roles)
                .claim("expiry", expiry.toInstant().toString())
                .id(UUID.randomUUID().toString())
                .issuedAt(now)
                .expiration(expiry)
                .signWith(signingKey)
                .compact();
    }

    // -----------------------------------------------------------------------
    // Token renewal (Task 1 AC2)
    // -----------------------------------------------------------------------

    /**
     * Renew an access token, even if it has expired.
     *
     * <p>Parses the token ignoring expiry, extracts the identity claims, and
     * issues a fresh token with a new expiry.
     *
     * @param token the (possibly expired) JWT string
     * @return a new signed JWT with refreshed expiry
     * @throws JwtValidationException if the token cannot be parsed at all
     */
    @SuppressWarnings("unchecked")
    public String renewToken(String token) {
        Claims claims;
        try {
            claims = Jwts.parser()
                    .verifyWith((javax.crypto.SecretKey) signingKey)
                    .clockSkewSeconds(Long.MAX_VALUE / 1000)
                    .build()
                    .parseSignedClaims(token)
                    .getPayload();
        } catch (JwtException e) {
            throw new JwtValidationException("Cannot renew invalid token: " + e.getMessage(), e);
        }

        String userId = (String) claims.get("userId");
        List<String> roles = (List<String>) claims.get("roles");
        if (userId == null) {
            throw new JwtValidationException("Token payload missing required userId claim");
        }

        return generateToken(userId, roles != null ? roles : List.of());
    }

    // -----------------------------------------------------------------------
    // Token validation (Task 1 AC3)
    // -----------------------------------------------------------------------

    /**
     * Validate a JWT token and verify the user has the required role.
     *
     * <p>Rejects tokens that:
     * <ul>
     *   <li>are expired</li>
     *   <li>have an invalid signature</li>
     *   <li>are missing required payload fields ({@code userId}, {@code roles})</li>
     *   <li>do not contain the {@code requiredRole}</li>
     * </ul>
     *
     * @param token        the JWT string to validate
     * @param requiredRole role that must be present in the token's roles claim
     * @return decoded {@link Claims} if valid
     * @throws JwtValidationException with a 401-semantic message on any failure
     */
    @SuppressWarnings("unchecked")
    public Claims validateToken(String token, String requiredRole) {
        Claims claims = parseAndValidateClaims(token);

        List<String> roles = (List<String>) claims.get("roles");
        if (roles == null || !roles.contains(requiredRole)) {
            throw new JwtValidationException(
                    "401: User does not have required role: " + requiredRole);
        }

        return claims;
    }

    /**
     * Validate a JWT token without role-checking.
     *
     * @param token the JWT string to validate
     * @return decoded {@link Claims} if valid
     * @throws JwtValidationException on any validation failure
     */
    public Claims validateToken(String token) {
        return parseAndValidateClaims(token);
    }

    // -----------------------------------------------------------------------
    // Private helpers
    // -----------------------------------------------------------------------

    private Claims parseAndValidateClaims(String token) {
        if (token == null || token.isBlank()) {
            throw new JwtValidationException("401: Token is missing or empty");
        }

        Claims claims;
        try {
            claims = Jwts.parser()
                    .verifyWith((javax.crypto.SecretKey) signingKey)
                    .build()
                    .parseSignedClaims(token)
                    .getPayload();
        } catch (ExpiredJwtException e) {
            throw new JwtValidationException("401: Token has expired", e);
        } catch (JwtException e) {
            throw new JwtValidationException("401: Invalid token: " + e.getMessage(), e);
        }

        // Reject tokens with incorrect payload formats (Task 1 AC3)
        if (claims.get("userId") == null) {
            throw new JwtValidationException("401: Token payload missing required userId claim");
        }

        return claims;
    }

    // -----------------------------------------------------------------------
    // Exception type
    // -----------------------------------------------------------------------

    /**
     * Thrown when JWT validation fails (maps to HTTP 401 Unauthorized).
     */
    public static class JwtValidationException extends RuntimeException {
        public JwtValidationException(String message) {
            super(message);
        }

        public JwtValidationException(String message, Throwable cause) {
            super(message, cause);
        }
    }
}
